src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt

```diff
@@ -47,8 +43,12 @@ class RsImplIndex : AbstractStubIndex<RustTypeFingerprint, RsImplItem>() {
                 RsImplItem::class.java
             ).asSequence()
             return when {
-                else -> elements.filter {
-                    canUnify(target, it.typeReference?.type ?: RustUnknownType)
+                else -> elements.filter { impl ->
+                    val ty = impl.typeReference?.type
+                    // Addition class check is a temporal solution to filter impls for type parameter
+                    // with the same name
+                    // struct S; impl<S: Tr1> Tr2 for S {}
+                    ty != null && ty.javaClass == target.javaClass && ty.isAssignableFrom(target, project)
```

Comment:
This would not be a problem, but the code given in the comment leads to a StackOverflowError.